faunadb-java-dsl/src/main/java/com/faunadb/client/types/Decoder.java

```diff
@@ -0,0 +1,309 @@
+package com.faunadb.client.types;
+
+import com.faunadb.client.errors.FaunaException;
+import com.faunadb.client.types.Creators.Creator;
+import com.faunadb.client.types.Types.CollectionType;
+import com.faunadb.client.types.Types.SimpleType;
+import com.faunadb.client.types.Types.MapType;
+import com.faunadb.client.types.Value.*;
+import com.faunadb.client.types.time.HighPrecisionTime;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import org.joda.time.Instant;
+import org.joda.time.LocalDate;
+
+import java.lang.reflect.*;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.google.common.collect.Maps.newHashMap;
+import static java.lang.String.format;
+
+/**
+ * FaunaDB {@link Value} to object decoder.
+ */
+public class Decoder {
```

Comment:
I've been thinking about the mechanics here. I think it would be more fluid for users if we manage to use the `Value` interface for conversion in the same way we do today.

Something more like:
```java
User inst = aValue.to(User.class).get();
```